File: classes-years/2021/input_devices/mag/vector/hello.ALS31313.t412.py
#
# hello.ALS31313.t412.py
#    vector magnetometer tiny412 hello-world
#       hello.ALS31313.t412.py serial_port
#
# Neil Gershenfeld 12/19/21
#
# This work may be reproduced, modified, distributed,
# performed, and displayed for any purpose. Copyright is
# retained and must be preserved. The work is provided
# as is; no warranty is provided, and users accept all 
# liability.
#

from tkinter import *
import serial

WINDOW = 600 # window size
nsamples = 100.0 # number of samples accumulated
xf = yf = zf = 0 # filtered values
eps = 0.01 # filter fraction
scale = 200 # bar graph scale

def idle(parent,canvas):
   global xf,yf,zf
   #
   # idle routine
   #
   byte2 = 0
   byte3 = 0
   byte4 = 0
   ser.flush()
   while 1:
      #
      # find framing 
      #
      byte1 = byte2
      byte2 = byte3
      byte3 = byte4
      byte4 = ord(ser.read())
      if ((byte1 == 1) & (byte2 == 2) & (byte3 == 3) & (byte4 == 4)):
         break
   x0 = ord(ser.read())
   y0 = ord(ser.read())
   z0 = ord(ser.read())
   x1 = ord(ser.read())
   y1 = ord(ser.read())
   z1 = ord(ser.read())
   x = (x0 << 4) + (x1 & 0b1111)
   y = (y0 << 4) + (y1 & 0b1111)
   z = (z0 << 4) + (z1 & 0b1111)
   if (x > 2047):
      x = x-4096
   if (y > 2047):
      y = y-4096
   if (z > 2047):
      z = z-4096
   xf = (1-eps)*xf+eps*x
   yf = (1-eps)*yf+eps*y
   zf = (1-eps)*zf+eps*z
   canvas.itemconfigure("xtext",text="%.1f"%xf)
   canvas.itemconfigure("ytext",text="%.1f"%yf)
   canvas.itemconfigure("ztext",text="%.1f"%zf)
   xp = (0.2+(0.9-0.2)/2.0)*WINDOW+((0.9-0.2)/2)*WINDOW*xf/scale
   yp = (0.2+(0.9-0.2)/2.0)*WINDOW+((0.9-0.2)/2)*WINDOW*yf/scale
   zp = (0.2+(0.9-0.2)/2.0)*WINDOW+((0.9-0.2)/2)*WINDOW*zf/scale
   canvas.coords('xrect1',.2*WINDOW,.05*WINDOW,xp,.2*WINDOW)
   canvas.coords('xrect2',xp,.05*WINDOW,.9*WINDOW,.2*WINDOW)
   canvas.coords('yrect1',.2*WINDOW,.25*WINDOW,yp,.4*WINDOW)
   canvas.coords('yrect2',yp,.25*WINDOW,.9*WINDOW,.4*WINDOW)
   canvas.coords('zrect1',.2*WINDOW,.45*WINDOW,zp,.6*WINDOW)
   canvas.coords('zrect2',zp,.45*WINDOW,.9*WINDOW,.6*WINDOW)
   canvas.update()
   parent.after_idle(idle,parent,canvas)

#
#  check command line arguments
#
if (len(sys.argv) != 2):
   print("command line: hello.ALS31313.t412.py serial_port")
   sys.exit()
port = sys.argv[1]
#
# open serial port
#
ser = serial.Serial(port,115200)
ser.setDTR()
#
# set up GUI
#
root = Tk()
root.title('hello.ALS31313.t412.py (q to exit)')
root.bind('q','exit')
canvas = Canvas(root, width=WINDOW, height=.65*WINDOW, background='white')
#
canvas.create_text(.1*WINDOW,.125*WINDOW,text="",font=("Helvetica", 24),tags="xtext",fill="#0000b0")
canvas.create_text(.95*WINDOW,.125*WINDOW,text="x",font=("Helvetica", 24),fill="#0000b0")
canvas.create_rectangle(.2*WINDOW,.05*WINDOW,.3*WINDOW,.2*WINDOW, tags='xrect1', fill='#b00000')
canvas.create_rectangle(.3*WINDOW,.05*WINDOW,.9*WINDOW,.2*WINDOW, tags='xrect2', fill='#0000b0')
#
canvas.create_text(.1*WINDOW,.325*WINDOW,text="",font=("Helvetica", 24),tags="ytext",fill="#0000b0")
canvas.create_text(.95*WINDOW,.325*WINDOW,text="y",font=("Helvetica", 24),fill="#0000b0")
canvas.create_rectangle(.2*WINDOW,.25*WINDOW,.3*WINDOW,.4*WINDOW, tags='yrect1', fill='#b00000')
canvas.create_rectangle(.3*WINDOW,.25*WINDOW,.9*WINDOW,.4*WINDOW, tags='yrect2', fill='#0000b0')
#
canvas.create_text(.1*WINDOW,.525*WINDOW,text="",font=("Helvetica", 24),tags="ztext",fill="#0000b0")
canvas.create_text(.95*WINDOW,.525*WINDOW,text="z",font=("Helvetica", 24),fill="#0000b0")
canvas.create_rectangle(.2*WINDOW,.45*WINDOW,.3*WINDOW,.6*WINDOW, tags='zrect1', fill='#b00000')
canvas.create_rectangle(.3*WINDOW,.45*WINDOW,.9*WINDOW,.6*WINDOW, tags='zrect2', fill='#0000b0')
canvas.pack()
#
# start idle loop
#
root.after(100,idle,root,canvas)
root.mainloop()

File: classes-years/2021/input_devices/mag/vector/hello.ALS31313.t412.c
//
// hello.ALS31313.t412.c
//    vector magnetometer tiny412 hello-world
//
// Neil Gershenfeld 12/19/21
//
// This work may be reproduced, modified, distributed,
// performed, and displayed for any purpose. Copyright is
// retained and must be preserved. The work is provided
// as is; no warranty is provided, and users accept all 
// liability.
//

#include <avr/io.h>
#include <util/delay.h>

#define output(directions,pin) (directions |= pin) // set port direction for output
#define input(directions,pin) (directions &= (~pin)) // set port direction for input
#define set(port,pin) (port |= pin) // set port pin
#define clear(port,pin) (port &= (~pin)) // clear port pin
#define bit_delay_time 8.5 // bit delay for 115200 with overhead
#define bit_delay() _delay_us(bit_delay_time) // RS232 bit delay
#define half_bit_delay() _delay_us(bit_delay_time/2) // RS232 half bit delay
#define pin_test(pins,pin) (pins & pin) // test for port pin
#define bit_test(byte,bit) (byte & (1 << bit)) // test for bit set

#define serial_port_direction VPORTA.DIR
#define serial_port_out VPORTA.OUT
#define serial_port_in VPORTA.IN
#define serial_pin_out PIN6_bm
#define serial_pin_in PIN7_bm

#define address 96

#define I2C_delay() _delay_us(10)
#define SCL_pin PIN2_bm
#define SCL_pins VPORTA.IN
#define SCL_port VPORTA.OUT
#define SCL_direction VPORTA.DIR
#define SDA_pin PIN1_bm
#define SDA_pins VPORTA.IN
#define SDA_port VPORTA.OUT
#define SDA_direction VPORTA.DIR

void SCL_write(char bit) {
   //
   // write SCL bit
   //
   if (bit == 0) {
      output(SCL_direction,SCL_pin);
      clear(SCL_port,SCL_pin);
      }
   else {
      input(SCL_direction,SCL_pin);
      while (pin_test(SCL_pins,SCL_pin) == 0); // check for clock stretching
      }
   }

void SDA_write(char bit) {
   //
   // write SDA bit
   //
   if (bit == 0) {
      output(SDA_direction,SDA_pin);
      clear(SDA_port,SDA_pin);
      }
   else
      input(SDA_direction,SDA_pin);
   }

void I2C_init() {
   //
   // initialize I2C lines
   //
   SDA_write(1);
   SCL_write(1);
   } 

char I2C_master_write_byte(unsigned char byte) {
   //
   // master write I2C byte
   //
   unsigned char bit;
   //
   // loop over bits
   //
   for (bit = 0; bit < 8; ++bit) {
      if ((byte & 0x80) == 0)
         SDA_write(0);
      else
         SDA_write(1);
      SCL_write(1);
      I2C_delay();
      SCL_write(0);
      I2C_delay();
      byte <<= 1;
      }
   //
   // check for ACK
   //
   SDA_write(1);
   SCL_write(1);
   I2C_delay();
   if (pin_test(SDA_pins,SDA_pin) != 0) {
      //
      // no ACK, return 1
      //
      ;//return 2;	
      }
   //
   // yes ACK, return 0
   //
   SCL_write(0);
   I2C_delay();
   return 0;
   }	

char I2C_master_write(unsigned char* data, unsigned char nbytes, unsigned char secondary_address) {
   //
   // I2C master write
   //
   unsigned char index,ret,secondary_address_write;
   //
   // send start
   //
   SDA_write(0);
   I2C_delay();
   SCL_write(0);	
   I2C_delay();
   //
   // send secondary address
   //
   secondary_address_write = secondary_address << 1;
   if (I2C_master_write_byte(secondary_address_write) != 0)
      //
      // no ACK, return 1
      //
      ;//return 3;
   //
   // loop over bytes
   //
   for (index = 0; index < nbytes; ++index) {
      ret = I2C_master_write_byte(data[index]);
      if (ret != 0)
         //
         // no ACK, return 1
         //
         ;//break;
      //
      // yes ACK, continue
      //	
      }
   //
   // send stop
   //
   SCL_write(1);
   I2C_delay();
   SDA_write(1);
   I2C_delay();
   return ret;
   }

void I2C_master_read_byte(unsigned char* data, unsigned char index, unsigned char nbytes) {
   //
   // master read I2C byte
   //
   unsigned char byte,bit;
   SDA_write(1);
   byte = 0;
   //
   // loop over bits
   //
   for (bit = 0; bit < 8; ++bit)  {
      SCL_write(1);
      I2C_delay();
      if (pin_test(SDA_pins,SDA_pin) != 0)
         byte |= (1 << (7-bit));
      SCL_write(0);
      I2C_delay();
      }
   data[index] = byte;
   if (index < (nbytes-1)) {
      //
      // not done, send ACK
      //
      SDA_write(0);
      SCL_write(1);
      I2C_delay();
      SCL_write(0);
      SDA_write(1);
      I2C_delay();
      }
   else {
      //
      // done, send NACK
      //
      SDA_write(1);
      SCL_write(1);
      I2C_delay();
      SCL_write(0);
      I2C_delay();
      }
   }	

char I2C_master_read(unsigned char* data, unsigned char nbytes, unsigned char secondary_address) {
   //
   // I2C master read
   //
   unsigned char index,secondary_address_read;
   //
   // send start
   //
   SDA_write(0);
   I2C_delay();
   SCL_write(0);	
   I2C_delay();
   //
   // send secondary address
   //
   secondary_address_read = (secondary_address << 1) + 1;
   if (I2C_master_write_byte(secondary_address_read) == 1)
      //
      // no ACK, return 1
      //
      return 4;
   //
   // loop over bytes
   //
   for (index = 0; index < nbytes; ++index)
      I2C_master_read_byte(data,index,nbytes);
   //
   // send stop
   //
   SCL_write(1);
   I2C_delay();
   SDA_write(1);
   I2C_delay();
   return 0;
   }	

void put_char(volatile unsigned char *port, unsigned char pin, char txchar) {
   //
   // send character in txchar on port pin
   //    assumes line driver (inverts bits)
   //
   // start bit
   //
   clear(*port,pin);
   bit_delay();
   //
   // unrolled loop to write data bits
   //
   if bit_test(txchar,0)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,1)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,2)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,3)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,4)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,5)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,6)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,7)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   //
   // stop bit
   //
   set(*port,pin);
   bit_delay();
   //
   // char delay
   //
   bit_delay();
   }

int main(void) {
   //
   // main
   //
   uint8_t ret,data[8];
   //
   // unprotect clock and turn off prescalar (20 MHz)
   //
   CPU_CCP = CCP_IOREG_gc;
   CLKCTRL.MCLKCTRLB = 0;
   //
   // initialize pins
   //
   set(serial_port_out,serial_pin_out);
   set(serial_port_direction,serial_pin_out);
   //
   // main loop
   //
   while (1) {
      //
      // read fields
      //
      data[0] = 0x28;
      ret = I2C_master_write(data,1,address);
      ret = I2C_master_read(data,8,address);
      //
      // send framing
      //
      put_char(&serial_port_out,serial_pin_out,1);
      put_char(&serial_port_out,serial_pin_out,2);
      put_char(&serial_port_out,serial_pin_out,3);
      put_char(&serial_port_out,serial_pin_out,4);
      //
      // send data
      //
      put_char(&serial_port_out,serial_pin_out,data[0]);
      put_char(&serial_port_out,serial_pin_out,data[1]);
      put_char(&serial_port_out,serial_pin_out,data[2]);
      put_char(&serial_port_out,serial_pin_out,data[5]);
      put_char(&serial_port_out,serial_pin_out,data[6]);
      put_char(&serial_port_out,serial_pin_out,data[7]);
      }
   }
